vignettes/design.Rmd

```diff
@@ -0,0 +1,210 @@
+* The `$state()` method returns the model state and accepts an argument `index` as the state to return
+
+In both cases, if `index` was named then the returned state carries these names as its rownames.
+
+The ordering of the state is important; we always have dimensions that will contain
+
+1. the states within a single particle
+2. the particles within a time (may be several dimensions; see `vignette("multi")`)
+3. the time dimension if using
+
+This is to minimise churn during writing, and to help with concatenation. Multiple particles data is stored consecutively and read and written in order. Each time step is written at once. And you can append states from different times easily. The `aperm()` function will be useful for reshaping this output to a different dimension order if you require one.
+
+In order to pull all of this off, we allocate all our memory up front, in C++ and pass back to R a "pointer" to this memory, which will live for as long as your model object. This means that even if your model requires GBs of memory to run, it is never copied back and forth into R (where it would be subject to R's [copy-on-write semantics](https://en.wikipedia.org/wiki/Copy-on-write) but instead accessed only when needed, and written to in place following C++ reference semantics.
+
+## Useful verbs
+
+We try and provide verbs that are useful, given that the model presents a largely opaque pointer to model state. These are driven by our needs for running a particle filter.
+
+Normally we have in the object several things:
+
+* the random number state: this is effectively a matrix of 64-bit integers
+* the model state: this is effectively a matrix of floating point numbers, typically `double`s
+* the model parameters: this is specific to the model in question and these are likely shared across multiple particles). This is presented to all particles within a group as an immutable pointer, which allows safe simultaneous access by multiple threads
+* the model _internal state_: this is also specific to the model and dust has no control over this. Because each particle has its own state they can safely write to it when running parallel (compare with parameters). Typically this space is *allocated* at model construction. For models that come from [odin.dust](https://mrc-ide.github.io/odin.dust/) this is used as "scratch" space that we write to during an update but which does not conceptually persist between steps.
+
+There is one subtlety about internal state: we assume that the state entirely specifies a model in a Markov process and we don't guarantee that models with mutable internal state that is not conceptually discarded each iteration. This is because if we reorder particles what we really do is reorder the *state vector* and not this internal state. This prevents implementing things like models with "delays" in the current design. We may relax constraint this if it is needed.
```

Comment:
```suggestion
There is one subtlety about internal state: we assume that the state entirely specifies a model in a Markov process, and so we don't guarantee that models with mutable internal state will not be discarded between each iteration. This is because if we reorder particles what we really do is reorder the *state vector* and not this internal state. This prevents implementing things like models with "delays" in the current design. We may relax constraint this if it is needed.
```